### ISSUE TITLE
pasting URL always generates full link syntax 

### ISSUE THREAD

**Issue by brainchild0** [marktext/marktext#2259]
# Issue

Suppose the user has written the following text:

`````
Elephants are the largest existing land animal (see [Wikipedia article](
`````

Then, pasting the URL `https://en.wikipedia.org/wiki/Elephant`, and then typing `)).`, should result in  text as follows: 

```
Elephants are the largest existing land animal (see [Wikipedia article](https://en.wikipedia.org/wiki/Elephant)).
```

In fact the result is the following:

```
Elephants are the largest existing land animal (see [Wikipedia article]([Elephant - Wikipedia](https://en.wikipedia.org/wiki/Elephant))).
```

# Environment

- **OS/Desktop:** Mint 19.3, Cinnamon 4.4.8
- **Application version:** 16.2

# Notes

Strangely, both display the same in the editing buffer, which may be a separate bug.

**Comment by jomNext** [comment]
This is not a fault, i like this function. If you insert a link in Marktext via copy-and-paste, the name (extract from the website) is automatically adopted. If you want to enter the name yourself, as in your example, you can insert the text as pure text using the shortcut Ctrl + Shift + T.

**Comment by brainchild0** [comment]
I have no problem with  both behaviors supported through different commands, but   this behavior occurring with the paste command is confusing.  A user who copies a URL from a browser window believes that the text in the clipboard is that URL, which would be pasted verbatim.

If the simple paste is already supported, perhaps the best solution is  simply reversing how the application resolves which action.  If regular pasting simply inserted the URL, then  *ctrl*-*shift*-*V* might create the formatted link.

In either case, the documented effect is certainly a bug.  Inserting a formatted link while part of the link syntax is in the buffer left of the caret will always create incorrect Markdown, and is always not the intention of the user. It might be appropriate to introduce a special case.

**Comment by brainchild0** [comment]
Loosely related to #2029.

**Comment by makeworld-the-better-one** [comment]
I would at least like a way to disable this feature, so I can go back to using Ctrl-V because I prefer it. Thank you for your work on Marktext.

**Comment by guillemglez** [comment]
I would suggest to at least allow for disabling this functionality. It is confusing to the user since this behaviour is not expected from a simple paste (Ctrl + V) action.

**Comment by brainchild0** [comment]
I think that simply remapping the keyboard shortcuts and command names would be  an optimal solution. I believe that this approach preserves both kinds of functionality, while avoiding the addition of an option into the user configuration.

**Comment by guillemglez** [comment]
I understand you suggest that Ctrl + V should behave as Ctrl + Shift + V does at the moment? It does make sense. Actually in my case I did not know about the Shift option until I found this issue; was very frustrated on the default behaviour.

In my opinion it would be neater to have an "Automatically format pasted URL to link syntax" option. Changing current behaviour may produce the same frustration to users who are already used at the current functionality.

**Comment by brainchild0** [comment]
I understand that backward compatibility is important in general, but in the current case, the proposed change is only to a simple user command, which any user rather easily may relearn. 

More relevant is that the software is still in a pre-release phase. If it were mature, with a large user base, then this kind of change might be very disruptive. Presently, resolving an optimal design may be more important that maintaining such rigorous continuity with the current one.

**Comment by guillemglez** [comment]
To clarify, would this be the preferred approach in case of a pull request? (mentioning in the conversation @fxha @MrHeer @Jocs)

- Invert the behavior of Ctrl + V and Ctrl + Shift + V

and presumably,

- Add an option "automatically format URL to link syntax" which would have the effect described above (and should it be activated by default?)

**Comment by makeworld-the-better-one** [comment]
Yes that would be good.

**Comment by brainchild0** [comment]
>   Add an option "automatically format URL to link syntax" which would have the effect described above (and should it be activated by default?)

It is not clear to me what this item means, with respect to the two paste behaviors mentioned prior. Are you suggesting that the inversion ("behavior described above") would be conditional on a configuration option (disabling that option then preserving the current overall behavior)?

**Comment by guillemglez** [comment]
> Are you suggesting that the inversion ("behavior described above") would be conditional on a configuration option (disabling that option then preserving the current overall behavior)?

Yes. A boolean option so that the user may choose which one they prefer. I am not sure if it should be activated by default, though.

**Comment by brainchild0** [comment]
Addition of user options, I think, is preferably limited to those options that noticeably improve usability. I am having trouble imagining the case for why this option has such a helpful effect. What is inadequate, in your view, about simply switching the mapping?

**Comment by guillemglez** [comment]
Switching the mapping is the preferred behavior to me, I just want to make sure this is the preferred approach :)

**Comment by brainchild0** [comment]
I am trying to understand how it might benefit the application to make the behavior configurable rather than static?

**Comment by brainchild0** [comment]
Actually, after reviewing the comments in #2314, though I appreciate the effort for making a contribution, I find that I am confused broadly by the distinction of *Paste* versus *Paste as Markdown*. One question is what  these terms intend to mean, and another is whether the particular language is the best for explaining the intended meaning. A further question is whether this particular distinction actually relates directly to the particular distinction of pasting a formatted link versus simply a URI.

**Comment by guillemglez** [comment]
Core thing in the _Paste as Markdown_ thing is that I find it makes more sense having next to each other _Copy as Markdown_ and _Paste as Markdown_ (as Ctrl+Shift+[C/V]) rather than _Copy as Markdown_ and _Paste as Plaintext_. See below:

![image](https://user-images.githubusercontent.com/25482726/91408432-271f3c00-e844-11ea-8aef-24458a6b4897.png)

**Comment by brainchild0** [comment]
How would you explain the following five functions, currently supported?:

- *Paste*
- *Paste as Plain Text*
- *Copy*
- *Copy as Markdown*
- *Copy as HTML*

**Comment by guillemglez** [comment]
As I understood, the intention is:

- _Paste_: Will paste whatever is in the clipboard in such a way that it looks how it was intended in the rendered Markdown (processing takes place to convert it to Markdown source)

- _Paste as Plain Text_: Will paste whatever is in the clipboard as-is into the Markdown source

- _Copy_: Copies the text as formatted, no Markdown syntax is included

- _Copy as Markdown_: Copies selection from Markdown source

- _Copy as HTML_: Copies selection as HTML source

As we discussed here, in _Paste_ the plain text approach is to be prioritized. What I suggested is reorganizing these functionalities as:

- _Paste_: Will paste whatever is in the clipboard as-is into the Markdown source

- _Paste as Markdown_: Will paste whatever is in the clipboard in such a way that it looks how it was intended in the rendered Markdown (processing takes place to convert it to Markdown source)

- _Copy_: Copies the text as formatted, no Markdown syntax is included

- _Copy as Markdown_: Copies selection from Markdown source

- _Copy as HTML_: Copies selection as HTML source

I think this would make more sense towards the user since the copy and paste functionalities would be more along (see screenshot in my comment above).

(If this makes sense I think I will add this into the PR description, it'll make it easier to understand)

**Comment by brainchild0** [comment]
I find both the current and proposed language confusing.  I suggest that conversion and pasting from  a non-Markdown source in the clipboard should be called *Paste*, and pasting Markdown source verbatim into the buffer should be called *Paste from Markdown* (note *from* not *as*).

I also find it confusing how this distinction is related to whether  a URI is pasted as a formatted link from the page title, or rather just the address.

**Comment by guillemglez** [comment]
Yes... Also thought about this last point... For example, now let's say you have a formatted link `[link](http://link.com)` copied into the clipboard. What should the behavior be for each mode?

**Comment by brainchild0** [comment]
I think an even more basic question needs to be answered first. What is the behavior if the text in the clipboard is simply a URL...?:
1. ...in the regular case.
1. ...in the case that the text preceding the caret is `[link](`?

Currently, the behavior is to fetch the title of the web page from the network, and then create a formatted link, inserting it at the position of the caret.  My view, at the time of opening the issue as well as now, is that the behavior is probably undesirable generally for case (1), but certainly undesirable for case (2).

I personally see no substantial, practical benefit to fetching a page title, in part because that title is rarely the preferred link text in practice. If enough users find it useful, however, I have no problem supporting  an additional feature, separate from the basic (or any) paste functionality.

**Comment by makeworld-the-better-one** [comment]
@guillemglez @brainchild0 

I'd really just prefer if the text I paste (with Ctrl-V) is pasted irregardless of context or the content of the paste. It shouldn't be modified or affect other things. If you want to have context-aware things, that could be under Ctrl-Shift-V, so the user has to deliberately choose it.

**Comment by guillemglez** [comment]
I think that everyone agrees that `[<fetched title>](<link in clipboard>)` is not the desirable behavior. Other discussions with how the paste functionality should behave I think they should take place somewhere else.

Would you agree on having `[<link in clipboard>](<link in clipboard>)` default behavior instead of`[<fetched title>](<link in clipboard>)` in the context of this PR?

As well, answering @makeworld-the-better-one 

> I'd really just prefer if the text I paste (with Ctrl-V) is pasted irregardless of context or the content of the paste. It shouldn't be modified or affect other things. If you want to have context-aware things, that could be under Ctrl-Shift-V, so the user has to deliberately choose it.

I feel like this is a contextual editor (since you can write both source code and edit in a WYSIWYG manner) and that context is the most important parameter to define this behavior. From what @brainchild0 pointed out, I think it should not behave the same way in WYSIWYG mode than in source editing mode.

**Comment by brainchild0** [comment]
@guillemglez: I am having a hard time resolving which parts of this conversation may have been confused by multiple issues having been discussed as though they are one.

The original description concerned a specific case, that in which the clipboard contains a URL as plain text, as would be copied from a  browser address bar.

In the case that the clipboard contains a formatted link with a title and address, as copied from a browser content pane, and that  the paste mode is one in which formatting generally, such as emphasis, is preserved, then I would suggest that the link format similarly should be  preserved. I believe that such is the current behavior. Suggestions for changing it for some stated reason are welcome, but I believe now that the discussion would be clearer if such a suggestion appeared in a separate topic.

**Comment by guillemglez** [comment]
@brainchild0 To me the current behavior generates something the user hardly ever looks for. 

With a link in the clipboard, the expected output for _paste as markdown_ should be `[<link in clipboard>](<link in clipboard>)` in my opinion. Fetching the page title to place `[<page title>](<link in clipboard>)` is a wrong approach from my perspective.

To paste just the text with no link structure, _paste as plain text_ option should be used. 

Whether one paste mode or the other is default this I agree is a discussion which belongs elsewhere. The debate on the context (e.g. the user started typing link syntax) is also very necessary but does neither belong here.

**Comment by brainchild0** [comment]
@guillemglez: I agree with you largely, but make two relevant notes:

1. I believe that if the text in the clipboard is simply a plain URI, then the effect of a basic, and probably any, paste operation, should be to insert simply the URI. Creating a Markdown link in such cases appears to me to be undesirable.
1. The case of the clipboard containing simply a plain URI is the case that prompted my opening this issue. A special case is that in which the caret follows an unfinished instance of the link syntax. I am open to expanding the scope of this issue, as well as to anyone's opening further issues, but this particular range of cases is core to this particular issue.

**Comment by Jocs** [comment]
Paste will parse the style by default. For example, if it is a url, it will automatically generate link syntax. Can `paste as plain text` solve your problem?

**Comment by brainchild0** [comment]
The idea of the request is that in some cases, the input caret would be in a context  making formatting text an invalid insertion for parsing.

Reproducing the following original example, the application should detect that the input must be a plain URL:

```
Elephants are the largest existing land animal (see [Wikipedia article](
```

If the clipboard contents is simply a valid URL (e.g. `https://en.wikipedia.org/wiki/Elephant`), then the inserted text from a paste operation should be the same as inserting the URL by manual typing it from the keyboard.

**Comment by Anton-V-K** [comment]
I find the described behavior very frustrating.
I always forgot that pasting a link in MarkText produces extra markup in all cases, because it's not a typical behavior and certainly unexpected. In these cases reverting the undesired change requires manual editing and occasionally I even have to save the document with broken formatting to fix it in an external editog, because the cursor in MarkText jumps here and there while editing a bad-formatted link.
Please, fix it along with #2029 ASAP.

**Comment by Enashka** [comment]
I am really hating that pasting a URL inserts the article title as well. That's never what I mean to do.

**Comment by livejamie** [comment]
Copy/pasting my Request from #4959 here

---

### Summary

I use Markdown for speed and the control it gives me over the text I am inputting. When Marktext incorrectly parses links I wish I could change the default behavior for the document I'm working on, or even globally so I can Ctrl + V and keep going.

### Problem it solves

When working with websites that require a login and Marktext tries to get a preview of the URL it just provides a broken URL that I have to manually edit after pasting or I have to right click and click "paste as plain text" every single time.

When pasting a Reddit URL into Marktext this is what happens

<img width="395" height="78" alt="Image" src="https://github.com/user-attachments/assets/6687b9e2-59d6-4b15-97c7-6985dc1909ff" />

### Proposed solution

Allow users to toggle between rich text and plain text when pasting. Either as a togglable menu item or something within preferences.
